functions.sql

```diff
@@ -0,0 +1,86 @@
+/* Additional database functions for openstreetmap-carto */
+
+/* Access functions below adapted from https://github.com/imagico/osm-carto-alternative-colors/tree/591c861112b4e5d44badd108f4cd1409146bca0b/sql/roads.sql */
+
+/* Simplified 'yes', 'destination', 'no', 'unknown', NULL scale for access restriction 
+  'no' is returned if the rendering for highway category does not support 'restricted'.
+  NULL is functionally equivalent to 'yes', but indicates the absence of a restriction 
+  rather than a positive access = yes. 'unknown' corresponds to an uninterpretable 
+  access restriction e.g. access=unknown or motorcar=occasionally  */
+CREATE OR REPLACE FUNCTION carto_int_access(int_highway text, accesstag text)
+	RETURNS text
+	LANGUAGE SQL
+	IMMUTABLE PARALLEL SAFE
+AS $$
+SELECT
+	CASE
+	WHEN accesstag IN ('yes', 'designated', 'permissive') THEN 'yes'
+	WHEN accesstag IN ('destination',  'delivery', 'customers') THEN
+		CASE WHEN int_highway IN ('road', 'pedestrian') THEN 'restricted' ELSE 'yes' END
+	WHEN accesstag IN ('no', 'permit', 'private', 'agricultural', 'forestry', 'agricultural;forestry') THEN 'no'
+	WHEN accesstag IS NULL THEN NULL
+	ELSE 'unknown'
+	END
+$$;
+
+/* Try to promote path to cycleway (if bicycle allowed), then bridleway (if horse)
+   This duplicates existing behaviour where designated access is required */
+CREATE OR REPLACE FUNCTION carto_path_type(bicycle text, horse text)
+	RETURNS text
+	LANGUAGE SQL
+	IMMUTABLE PARALLEL SAFE
+AS $$
+SELECT
+	CASE
+		WHEN bicycle IN ('designated') THEN 'cycleway'
+		WHEN horse IN ('designated') THEN 'bridleway'
+		ELSE 'path'
+	END
+$$;
+
+/* Coalesce highways that will be treated in the same way, e.g. all roads become 'road'
+   Note that bicycle, horse arguments are only relevant if considering highway=path */
+CREATE OR REPLACE FUNCTION carto_highway_int_highway(highway text, bicycle text, horse text)
```

Comment:
This SQL function is called in exactly one place and consists of a single CASE statement. This makes it equivalent to
```sql
SELECT
CASE CASE WHEN highway in (..) THEN 'road' WHEN ... END
WHEN 'road' THEN ... END
WHEN 'pedestrian' THEN ... END
END
```

 Just modify the SQL statement calling it to remove the need for it.

Reply:
It's not quite that simple since `int_highway` is passed through to `carto_int_access`. But I would propose splitting `carto_int_access` into 2 vs 3 state functions, removing the need for `int_highway`.